js-pkg/client/src/provider.ts

```diff
@@ -210,443 +50,392 @@ type WebSocketPolyfillType = {
+        this.updateAwareness(decoder)
+        break
+      case messageQueryAwareness:
+        this.queryAwareness()
+        break
+      default:
+        break
     }
   }
 
-  connect() {
-    this.shouldConnect = true
-    if (!this.wsconnected && this.ws === null) {
-      setupWS(this)
-      this.connectBc()
+  private websocketClose(event: CloseEvent) {
+    this.emit('connection-close', event)
+    this.setSynced(false)
+    this.setStatus({ status: STATUS_DISCONNECTED })
+
+    // Remove all awareness states except for our own.
+    awarenessProtocol.removeAwarenessStates(
+      this.awareness,
+      Array.from(this.awareness.getStates().keys()).filter(
+        (client) => client !== this.doc.clientID,
+      ),
+      this,
+    )
+  }
+
+  private broadcastAwareness() {
+    if (this.awareness.getLocalState() !== null) {
+      const encoderAwarenessState = encoding.createEncoder()
+      encoding.writeVarUint(encoderAwarenessState, messageAwareness)
+      encoding.writeVarUint8Array(
+        encoderAwarenessState,
+        awarenessProtocol.encodeAwarenessUpdate(this.awareness, [this.doc.clientID]),
+      )
+      this.websocket?.send(encoding.toUint8Array(encoderAwarenessState))
     }
   }
 
-  addOnFailureHandler(handler: () => void): () => void {
-    this.onFailureHandlers.push(handler)
-    return () => {
-      this.onFailureHandlers = this.onFailureHandlers.filter((h) => h !== handler)
+  private websocketError(event: Event) {
+    this.emit('connection-error', event)
+  }
+
+  protected emit(eventName: YSweetEvent, data: any = null): void {
+    const listeners = this.listeners.get(eventName)
+    if (listeners) {
+      for (const listener of listeners) {
+        listener(data)
+      }
     }
   }
-}
 
-async function getClientToken(authEndpoint: AuthEndpoint, roomname: string): Promise<ClientToken> {
-  if (typeof authEndpoint === 'function') {
-    return await authEndpoint()
+  private setSynced(state: boolean) {
+    if (this.synced !== state) {
+      this.synced = state
+      this.emit('sync', state)
+    }
   }
-  const body = JSON.stringify({ docId: roomname })
-  const res = await fetch(authEndpoint, {
-    method: 'POST',
-    body,
-    headers: { 'Content-Type': 'application/json' },
-  })
-  // TODO: handle errors
-  const clientToken = await res.json()
-  // TODO: check that clientToken.docId === this.roomname
-  return clientToken
-}
 
-function updateProviderParams(
-  providerParams: YSweetProviderParams,
-  clientToken: ClientToken,
-): YSweetProviderParams {
-  // if the clientToken has a token, add it to the provider params
-  const connectParams = providerParams.params ? { ...providerParams.params } : {}
-  if (clientToken.token) connectParams.token = clientToken.token
-  return { ...providerParams, params: connectParams }
-}
+  private setStatus(status: YSweetStatus) {
+    if (status.status === STATUS_DISCONNECTED && this.shouldConnect) {
+      this.connect()
+    }
```

Comment:
This feels like it would be more appropriate in `websocketClose` (bonus, we can ditch the `status.status === STATUS_DISCONNECTED` check)